src/gsi/hybrid_ensemble_isotropic.F90

```diff
@@ -4533,7 +4723,13 @@ subroutine hybens_localization_setup
    ! but will need to rearrange so this can be set in control_vectors
    ! and triggered by lsqrtb.
    if ( regional ) then
-      nval_lenz_en = grd_loc%nlat*grd_loc%nlon*(grd_loc%kend_alloc-grd_loc%kbegin_loc+1)
+      if ( l_mgbf_loc ) then
```

Comment:
here, no need to have separate calculation for mgbf and RF, they have to be equal.

Reply:
nval_lenz_en is the number of components of hwork for RF, but VALL for MGBF. Since VALL is different from hwork, nval_lenz_en also should be different between RF and MGBF. Could you confirm how to use nval_lenz_en in subroutines ckgcov_a_en_new_factorization{,_ad}?

Reply:
nval_lenz_en is the size of (nlat*nlon* nk), where nlat and lon are the global dimension in horizontal directions and nk is the number of the vertical levels of the fields in a_ens distributed in this mpi task.  So, they should be the same no matter if or not use mbgf.   Or, it should be made sure "the same " that they could be used by mgbf.   If they are different, that means something wrong.

Reply:
As you pointed out, nval_lenz_en is the number of grid points in each processors, but it is different between vertically-parallelized RF and horizontally-parallelized MGBF. Please note that the domain of each processor is different between RF and MGBF. In horizontally-parallelized MGBF, this number of grid points in each processor is horizontally-local and vertically-global. That's why nval_lenz_en should be set as in this code in MGBF. There is no problem with this difference because the components of z(1:nval_lenz_en) don't affect each other between adjoint and direct filters.

Reply:
@shoyokota , On top of our above discussion on this question, now  I think the current specification of nval_lenz_en at Line 4789 is a mistake.  Because as mentioned earlier, because the results of rhs give the size of a local array like work_mbgf , while the nval_lenz_en is for the vertically split group of horizontally global arrays like hwork.  
Could you re-run your mgbf case to print out values from the two equations?  (better  use debug mode compiled GSI).

Reply:
Accordingly, the specification of obj_mbgf like 
```
! Apply horizontal smoother for number of horizontal scales
     ii=0
     do k=1,obj_mgbf(ig)%km_all
        do j=1-obj_mgbf(ig)%hy,obj_mgbf(ig)%jm+obj_mgbf(ig)%hy
           do i=1-obj_mgbf(ig)%hx,obj_mgbf(ig)%im+obj_mgbf(ig)%hx
              ii=ii+1
              obj_mgbf(ig)%VALL(k,i,j)=z(ii)
           enddo
        enddo
     enddo
```,   is questionable.    We could have a off-line discussion on this.

Reply:
And, nval_lenz_en is used as a global dimension size parameter  beyond hybrid_ensemble_iostropic.f90.   I think the current definition in mgbf branch would be problematic .

Reply:
@TingLei-NOAA In MGBF, nval_lenz_en is the size of horizontally-local vertically-global array and different from the original horizontally-global vertically-local nval_lenz_en for RF, but it doesn't cause any problems as I explained above. Most important benefit of MGBF is not to require all-to-all communication to change the domain of each processor. To attain that, we should not define z(1 : nval_lenz_en) as a horizontally global array. Namely, we should change the definition of z(1 : nval_lenz_en).

I'm not sure about your suggestion. In case nval_lenz_en_mgbf is newly defined separately from nval_lenz_en, how should we define z(1 : nval_lenz_en)?

Reply:
I finally understand that you are warried that nval_lenz_en is used also outside of hybrid_ensemble_isotropic.F90. Except for this file, it looks used only in jfunc.f90 (only for lsqrtb=true). My test was not failed due to lsqrtb=false, but it may be failed in lsqrtb=true as you pointed out.

I modified to use nval_loc_en instead of nval_lenz_en in hybrid_ensemble_isotropic.F90, and to force l_mgbf_loc=false in lsqrtb=true in gsimod.F90.

Reply:
@shoyokota  happy to know we have better understanding on this.  I think nval_lenz_en is not only used for lsqrtb, like sub random_cv . but it is not used in regular runs.  and, maybe it would be run (even turned on), it would be done before nval_lenz_en would have been changed.   
however, it is always safe/good to make your change here be local as in the updated version.

Reply:
Thank you for confirming it. Just for clarification, nval_lenz_en in control_vectors.f90 (e.g., for random_cv) is a local parameter, which is input from setup_control_vectors. It is not the same as nval_lenz_en defined in hybrid_ensemble_isotropic.F90 unless lsqrtb=true.